source/common/common/thread.cc

```diff
@@ -13,53 +13,50 @@ namespace {
 // Singleton structure capturing which thread is the main dispatcher thread, and
 // which is the test thread. This info is used for assertions around catching
 // exceptions and accessing data structures which are not mutex-protected, and
-// are expected only from the main thread.
+// are expected to be used for debug mode checks.
 //
 // TODO(jmarantz): avoid the singleton and instead have this object owned
 // by the ThreadFactory. That will require plumbing the API::API into all
 // call-sites for isMainThread(), which might be a bit of work, but will make
 // tests more hermetic.
 struct ThreadIds {
   bool inMainThread() const {
-    // We don't take the lock when testing the thread IDs, as they are atomic,
-    // and are cleared when being released. All possible thread orderings
-    // result in the correct result even without a lock.
-    return std::this_thread::get_id() == main_thread_id_;
+    absl::MutexLock lock(&mutex_);
+    return main_threads_to_usage_count_.find(std::this_thread::get_id()) !=
```

Comment:
When I originally added this @htuch was concerned that a mutex-lock while checking, even under an ASSERT, might introduce serialization in tests run with ASSERTs enabled, and reduce the chance of hitting unrelated races.

But looking at the code I don't see this as a big problem -- the lock is only held during a single hash lookup.

OTOH it looks very easy to use a `thread_local` bool here to do the inMainThread check. Any reason not to do that?

Reply:
I'm wasn't worried about major contention blocking causing performance distortion (I think), but probably just differences in timing when there is some correlated access. If there is no easy way to do this and remote singleton, let's go with this PR, but if thread_local works I agree it might be a bit better.